internal/stats/count_map.go

```diff
@@ -0,0 +1,80 @@
+package stats
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// CountersMap is a concurrency-safe map from keys of type K to uint32 counters.
+// It allows increments and retrievals of counts from concurrent go routines
+// without additional concurrency coordination.
+// Added counters cannot be removed.
+type CountersMap[K comparable] struct {
+	// Stores map[K]*uint32
+	// The counter is stored as pointer so it can be updated with atomic operations.
+	data   sync.Map
+	length atomic.Uint32
+}
+
+// Increment increases the counter for the specified key by 1.
+// Returns true if the key already existed, false if it was newly created.
+func (m *CountersMap[K]) Increment(key K) bool {
+	return m.Add(key, 1)
+}
+
+// Add increases the counter for the specified key by the value of v.
+// Returns true if the key already existed, false if it was newly created.
+func (m *CountersMap[K]) Add(key K, v uint32) bool {
```

Comment:
Do we need any kind of overflow protection or handling?

Reply:
Maybe.
This would overflow at 4 billion. Of course, it could be overflown by specifying large enough values for `Add`.
A (short-term?) option here is to unexport `Add`, and add a comment
saying that if it is to be exported, then overflow protections should be added.

@redgoat650  WDYT?

Reply:
What would be a reasonable behavior on overflow?
Given that this is used for stats, it seems that a `panic` would not be good behavior. Does it warrant adding a return error to the error signature?